app/src/main/java/org/mozilla/fenix/components/metrics/AppLaunchTimeMeasurement.kt

```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.os.Process
+import android.os.SystemClock
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.launch
+import org.mozilla.fenix.perf.Stat
+
+/**
+ * Handles the logic of figuring out launch time for cold, warm and hot startup.
+ */
+class AppLaunchTimeMeasurement(private val stats: Stat = Stat()) {
+
+    // when user launched the time. i.e press app_icon, launched a link.
+    private var processLaunchTimeStampNanoSeconds: Long? = null
+    // we are considering screen to be visible after the first pre draw call.
+    private var elapsedNanoTimestampOnHomeActivityPreDraw: Long? = null
+
+    fun onApplicationOnCreate() =
+        MainScope().launch(Dispatchers.IO) {
+        processLaunchTimeStampNanoSeconds = stats.getProcessStartTimeStampNano(Process.myPid())
+    }
+
+    fun onHomeActivityOnCreate(homeActivityInitTimeStampNanoSeconds: Long) {
+        if (processLaunchTimeStampNanoSeconds == null) {
+            // if startup timestamp was not set, we set the time here since the application might
+            // already exist and this is a warm startup
+            processLaunchTimeStampNanoSeconds = homeActivityInitTimeStampNanoSeconds
+        }
+    }
+
+    fun onHomeActivityOnRestart(currentTimeStampNanoSeconds: Long = SystemClock.elapsedRealtimeNanos()) {
+        processLaunchTimeStampNanoSeconds = currentTimeStampNanoSeconds
+    }
+
+    fun onActivityOnPreDraw(currentTimeNanoSeconds: Long = SystemClock.elapsedRealtimeNanos()) {
+        elapsedNanoTimestampOnHomeActivityPreDraw = currentTimeNanoSeconds
+    }
+
+    /**
+     * if we have both start and finish time for launch, return it otherwise we return null
+     */
+    fun getApplicationLaunchTime(): Long? {
+        return (elapsedNanoTimestampOnHomeActivityPreDraw?.minus(processLaunchTimeStampNanoSeconds!!))
```

Comment:
Why would elapsedNano every be null? Comment

Reply:
fixed